aspnetcore/blazor/security/blazor-web-app-with-windows-authentication.md

```diff
@@ -0,0 +1,137 @@
+
+```csharp
+builder.Services.AddAuthentication(NegotiateDefaults.AuthenticationScheme)
+    .AddNegotiate();
+```
+
+<xref:Microsoft.Extensions.DependencyInjection.PolicyServiceCollectionExtensions.AddAuthorization%2A> adds authorization policy services. <xref:Microsoft.AspNetCore.Authorization.AuthorizationOptions.FallbackPolicy%2A?displayProperty=nameWithType> sets the fallback authorization policy, which is set to the default policy (<xref:Microsoft.AspNetCore.Authorization.AuthorizationOptions.DefaultPolicy%2A?displayProperty=nameWithType>). The default policy requires an authenticated user to access the app:
+
+```csharp
+builder.Services.AddAuthorization(options =>
+{
+    options.FallbackPolicy = options.DefaultPolicy;
+});
+```
+
+<xref:Microsoft.Extensions.DependencyInjection.CascadingAuthenticationStateServiceCollectionExtensions.AddCascadingAuthenticationState%2A> adds cascading authentication state to the service collection. This is equivalent to placing a `CascadingAuthenticationState` component at the root of the app's component hierarchy:
+
+```csharp
+builder.Services.AddCascadingAuthenticationState();
+```
+
+An [authorization policy](xref:security/authorization/policies) is added for a [Windows security identifier (SID)](/windows-server/identity/ad-ds/manage/understand-security-identifiers). The `S-1-5-113` well-known SID in the following example indicates that the user is a local account, which restricts network sign-in to local accounts instead of "administrator" or equivalent accounts:
+
+```csharp
+builder.Services.AddAuthorizationBuilder()
+    .AddPolicy("LocalAccount", policy =>
+        policy.RequireClaim(
+            "http://schemas.microsoft.com/ws/2008/06/identity/claims/groupsid",
+            "S-1-5-113"));   
+```
+
+The authorization policy is enforced by the `LocalAccountOnly` component.
+
+`Components/Pages/LocalAccountOnly.razor`:
+
+```razor
+@page "/local-account-only"
+@using Microsoft.AspNetCore.Authorization
+@attribute [Authorize("LocalAccount")]
+
+<h1>Local Account Only</h1>
+
+<p>
+    You can only reach this page by satisfying the
+    <code>LocalAccount</code> authorization policy.
+</p>
+```
+
+The `UserClaims` component lists the user's claims, which includes the user's Windows security identifiers (SIDs).
+
+`Components/Pages/UserClaims.razor`:
+
+```razor
+@page "/user-claims"
+@using System.Security.Claims
+@using Microsoft.AspNetCore.Authorization
+@attribute [Authorize]
```

Comment:
What's the point of this attribute given that this sample also uses `options.FallbackPolicy = options.DefaultPolicy`?

I'd sooner remove the fallback policy part than this attribute though. I find too many people set the fallback policy to the default policy not realizing it makes `[Authorize]` redundant.

Reply:
I think the sample would need a `RedirectToLogin` like the OIDC sample has [here](https://github.com/dotnet/blazor-samples/blob/57e207bf450189f7fa1d5564a417712938732663/9.0/BlazorWebAppOidc/BlazorWebAppOidc.Client/Routes.razor#L5). It would probably best to call it something like `ForceLogin` even though, and I think you could get away with having the component just call `NavigationManager.Refresh(forceReload: true)` rather than hit the "authentication/login" minimal endpoint and redirect back, so it would be even simpler. The new top-level request induced by `Refresh` should be enough for the authorization middleware issue a challenge via the Windows authentication handler (`NegotiateHandler`).

Reply:
@halter73 ... I made the updates on https://github.com/dotnet/blazor-samples/pull/478/files.

Sample: https://github.com/dotnet/blazor-samples/tree/main/9.0/BlazorWebAppWinAuthServer

No 🎲🎲 yet ... Try to access a secure page, it presents the Windows Security signin popup, and then it puts the app/page into a redirect loop immediately after signing in.

Reply:
Ping @halter73 ... Made the changes that you recommended ☝️ ... but it puts the app into a redirect death loop.

Reply:
> No 🎲🎲 yet ... Try to access a secure page, it presents the Windows Security signin popup, and then it puts the app/page into a redirect loop of death 💀 immediately after signing in.

Does this happen even if you navigate directly to the secure page from the address bar? And does it happen after signing in successfully? I'm surprised that the `<NotAuthorized>` logic would reexecute after that.

I'll probably have to try the sample out myself to debug it. It might be a few days before I can get to it.

Reply:
> Does this happen even if you navigate directly to the secure page from the address bar?

You don't get a chance to do that under this scenario. The app loads the homepage with anonymous access. You navigate to a secure page ... sign in ... and then the authorized page loads, but it loads in a 💀 redirect loop, flashing it's content over and over rapidly.

> And does it happen after signing in successfully?

Yes, that's when it happens.